File: 프로그래머스/2/17684. ［3차］ 압축/［3차］ 압축.py
def solution(msg):
    answer = []
    T = "ABCDEFGHIJKLMNOPQRSTUVWXYZ"
    dic = {}
    for i in range(len(T)):
        dic[T[i]] = i+1
    
    l = len(msg)
    number = 26
    
    pass_cnt = 0
    i = 0
    while i != len(msg):
        pass_cnt = 0
        for j in range(i, l):
            if msg[i:j+1] not in dic.keys():
                number +=1
                dic[msg[i:j+1]] = number
                answer.append(dic[msg[i:j]])
                break
            else:
                if j+1 == l:
                    answer.append(dic[msg[i:j+1]])
                pass_cnt +=1
        i += 1 + pass_cnt -1
                
    return answer

File: 프로그래머스/2/17684. ［3차］ 압축/［3차］ 압축.js
function solution(msg) {
    let map = new Map();
    let answer = [];
    let index = 27;
    let w = "";
    let c = "";
    let temp = 0;
    
    for(let i = 0; i < msg.length; i++){
        w = msg[i];
        c = msg[i+1];
       if(!map.get(w+c)){
           map.set(w+c,index++);
           answer.push(w.charCodeAt()-64);
       }else{
           while(map.get(w+c)){
               temp = w+c;
               w = w+c;
               c = msg[++i+1];
           }
           map.set(w+c,index++);
           answer.push(map.get(temp));
       }
    }
    return answer;
}